### PR TITLE
docs(eslint-plugin): [restrict-template-expressions] add explanation for allowNumber

### LINKED ISSUE #9311: Document rationale for allowNumber: false in restrict-template-expressions
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22preset+config+change%22+) and found none that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Description

The strict-type-checked configuration sets `allowNumber: false` for the restrict-template-expressions rule. This prohibits code like the following:

```ts
const userId = 123;
const url = `/users/${userId}`;
```

There is nothing wrong with this code. The official [MDN docs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) on template literals contain code just like it. The [typescript-eslint docs](https://typescript-eslint.io/rules/restrict-template-expressions/#allownumber) for this rule offer no rationale of why interpolating numbers into a string in this way is problematic.

### Impacted Configurations

- strict
- strict-type-checked

### Additional Info

I'm totally bought into all the other options that `strict` enables for restrict-template-expressions (`allowAny`, `allowArray`, `allowBoolean`, `allowNullish`, `allowNever`). These seem potentially problematic since they could result in your inadvertently showing debug code to users. But `allowNumber`? I'm having trouble seeing any rationale for banning that.

My proposal would be to either change the default setting or add a clear explanation of why interpolating numbers into strings is dangerous.

Possibly related:
- #6758
- #5265

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
## PR Checklist

- [x] Addresses an existing open issue: fixes #9311
- [x] That issue was marked as [accepting prs](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aopen+is%3Aissue+label%3A%22accepting+prs%22)
- [x] Steps in [Contributing](https://typescript-eslint.io/contributing) were taken

## Overview

Adds a brief summary of reasoning as stated in #9311. Also copy & pastes one-liner descriptions for the rest of the options while I was in the area.

💖

### CODE REVIEW

File: packages/eslint-plugin/docs/rules/restrict-template-expressions.mdx [hunk 1]
```diff
@@ -65,8 +67,13 @@ const msg2 = `arg = ${arg || 'zero'}`;
 
 This option controls both numbers and BigInts.
 
+We recommend avoiding using this option if you use any floating point numbers.
```
Comment (Josh-Cena) [inline packages/eslint-plugin/docs/rules/restrict-template-expressions.mdx:70]:
```suggestion
We recommend using this option if you use any floating point numbers.
```

File: packages/eslint-plugin/docs/rules/restrict-template-expressions.mdx [hunk 2]
```diff
@@ -65,8 +67,13 @@ const msg2 = `arg = ${arg || 'zero'}`;
...
+Although `\`${1}\`` evaluates to `'1'`, `\`${0.1 + 0.2}\``evaluates to`'0.30000000000000004'`.
```
Comment (Josh-Cena) [inline packages/eslint-plugin/docs/rules/restrict-template-expressions.mdx:71]:
```suggestion
Although `` `${1}` `` evaluates to `'1'`, `` `${0.1 + 0.2}` ``evaluates to`'0.30000000000000004'`.
```

**Review by bradzacher — DISMISSED** [review]

File: packages/eslint-plugin/docs/rules/restrict-template-expressions.mdx (hunk 1 above)
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/docs/rules/restrict-template-expressions.mdx:70]: Err. The meaning right now is what I'd wanted to convey: if you use any floating point numbers, don't use this option. No?

Comment (Josh-Cena) [reply packages/eslint-plugin/docs/rules/restrict-template-expressions.mdx:70]: Ah, I get it. Yes.

File: packages/eslint-plugin/docs/rules/restrict-template-expressions.mdx [hunk 3]
```diff
@@ -65,8 +67,13 @@ const msg2 = `arg = ${arg || 'zero'}`;
 
 This option controls both numbers and BigInts.
 
+We recommend avoiding using this option if you use any floating point numbers.
+Although `` `${1}` `` evaluates to `'1'`, `` `${0.1 + 0.2}` `` evaluates to `'0.30000000000000004'`.
```
Comment (Josh-Cena) [inline packages/eslint-plugin/docs/rules/restrict-template-expressions.mdx:71]: Add an "instead..." Like "use a `toFixed` or `toPrecision` instead"

**Review by Josh-Cena — DISMISSED** [review]

File: packages/eslint-plugin/docs/rules/restrict-template-expressions.mdx [hunk 4]
```diff
@@ -65,8 +67,14 @@ const msg2 = `arg = ${arg || 'zero'}`;
...
+Consider using [`.toFixed()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toFixed) or [`toPrecision()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toPrecision) instead.
```
Comment (kirkwaiblinger) [inline packages/eslint-plugin/docs/rules/restrict-template-expressions.mdx:72]:
```suggestion
Consider using [`.toFixed()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toFixed) or [`.toPrecision()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toPrecision) instead.
```

**Review by kirkwaiblinger — APPROVED** [review]
pi

**Pull request merged**